contracts/protocol/v1/proxies/P1CurrencyConverterProxy.sol

```diff
@@ -0,0 +1,236 @@
+
+    // ============ Events ============
+
+    event LogConvertedDeposit(
+        address indexed account,
+        address perpetual,
+        address exchangeWrapper,
+        address tokenFrom,
+        address tokenTo,
+        uint256 tokenFromAmount,
+        uint256 tokenToAmount
+    );
+
+    event LogConvertedWithdrawal(
+        address indexed account,
+        address destination,
+        address perpetual,
+        address exchangeWrapper,
+        address tokenFrom,
+        address tokenTo,
+        uint256 tokenFromAmount,
+        uint256 tokenToAmount
+    );
+
+    // ============ State-Changing Functions ============
+
+    /**
+     * @notice Sets the maximum allowance on the Perpetual contract. Must be called at least once
+     *  on a given Perpetual before deposits can be made.
+     * @dev Cannot be run in the constructor due to technical restrictions in Solidity.
+     */
+    function approveMaximumOnPerpetual(
+        address perpetual
+    )
+        external
+    {
+        IERC20 tokenContract = IERC20(I_PerpetualV1(perpetual).getTokenContract());
+
+        // safeApprove requires unsetting the allowance first.
+        tokenContract.safeApprove(perpetual, 0);
+
+        // Set the allowance to the highest possible value.
+        tokenContract.safeApprove(perpetual, uint256(-1));
+    }
+
+    /**
+     * @notice Make a margin deposit to a Perpetual, after converting funds to the margin currency.
+     *  Funds will be withdrawn from the sender and deposited into the specified account.
+     * @dev Emits LogConvertedDeposit event.
+     *
+     * @param  account          The account for which to credit the deposit.
+     * @param  perpetual        The PerpetualV1 contract to deposit to.
+     * @param  exchangeWrapper  The ExchangeWrapper contract to trade with.
+     * @param  tokenFrom        The token to convert from.
+     * @param  tokenFromAmount  The amount of `tokenFrom` tokens to deposit.
+     * @param  data             Trade parameters for the ExchangeWrapper.
+     */
+    function deposit(
+        address account,
+        address perpetual,
+        address exchangeWrapper,
+        address tokenFrom,
+        uint256 tokenFromAmount,
+        bytes calldata data
+    )
+        external
+        returns (uint256)
+    {
+        I_PerpetualV1 perpetualContract = I_PerpetualV1(perpetual);
+        address tokenTo = perpetualContract.getTokenContract();
+        address self = address(this);
+
+        // Send fromToken to the ExchangeWrapper.
+        //
+        // TODO: Take possible ERC20 fee into account.
+        IERC20(tokenFrom).safeTransferFrom(
+            msg.sender,
+            exchangeWrapper,
+            tokenFromAmount
+        );
+
+        // Convert fromToken to toToken on the ExchangeWrapper.
+        I_ExchangeWrapper exchangeWrapperContract = I_ExchangeWrapper(exchangeWrapper);
+        uint256 tokenToAmount = exchangeWrapperContract.exchange(
+            self,
```

Comment:
I think technically this should be `msg.sender` but I don't expect this to be used by the exchangeWrapper anyway

Reply:
fixed